### DISCUSSION TITLE
What is DirichletBC().nodes ?

### DISCUSSION THREAD

**Discussion by APaganini** [firedrakeproject/firedrake#2451] [Firedrake support] [answered]
This is either a bug or I don't understand what `DirichletBC().nodes` means.
The output of the following code
```python
from firedrake import *
for N in [10, 50, 100]:  # various mesh refinements
    for p in range(1,6):  # various polynomial degrees 
        mesh = UnitIntervalMesh(N)
        V = FunctionSpace(mesh, 'CG', p)
        bcs = DirichletBC(V, 0.0, "on_boundary")
        print("N = %3u, p = %u, Vdim = %3u" % (N, p, V.dim()),
              "bcs.nodes = ", bcs.nodes)
```
is
```
N =  10, p = 1, Vdim =  11 bcs.nodes =  [ 0 10]
N =  10, p = 2, Vdim =  21 bcs.nodes =  [ 1 20]
N =  10, p = 3, Vdim =  31 bcs.nodes =  [ 2 30]
N =  10, p = 4, Vdim =  41 bcs.nodes =  [ 3 40]
N =  10, p = 5, Vdim =  51 bcs.nodes =  [ 4 50]
N =  50, p = 1, Vdim =  51 bcs.nodes =  [ 0 50]
N =  50, p = 2, Vdim = 101 bcs.nodes =  [  1 100]
N =  50, p = 3, Vdim = 151 bcs.nodes =  [  2 150]
N =  50, p = 4, Vdim = 201 bcs.nodes =  [  3 200]
N =  50, p = 5, Vdim = 251 bcs.nodes =  [  4 250]
N = 100, p = 1, Vdim = 101 bcs.nodes =  [  0 100]
N = 100, p = 2, Vdim = 201 bcs.nodes =  [  1 200]
N = 100, p = 3, Vdim = 301 bcs.nodes =  [  2 300]
N = 100, p = 4, Vdim = 401 bcs.nodes =  [  3 400]
N = 100, p = 5, Vdim = 501 bcs.nodes =  [  4 500]
```
After comparing the values of `V.dim()` and `bcs.nodes[-1]`, my guess is that `nodes` refers to the dofs associated with `DirichletBC` (which correspond to the mesh nodes for `CG1`elements). But why does `bcs.nodes[0]` change with the polynomial degree?

**Comment by ksagiyam** [comment, marked as answer]
Firedrake traverses the mesh entities (cells, facets, edges, vertices) and collects DoFs defined on each entity. The order of traversal is determined by `mesh._plex_renumbering` (https://github.com/firedrakeproject/firedrake/blob/e6e778ad4745d5d4ceee81722768c2b903ce6a11/firedrake/mesh.py#L826) (basically, a depth-first DAG traversal).

For instance, for N = 10 and p = 4, Firedrake collects DoFs as the following:

x-o-o-o-x-o-o-o-x...x-o-o-o-x
3 0 1 2 4 5 6 7 8 ... 36 37 38 39 40,

which agrees with your observation. We should note, however, that it is not guaranteed that we visit the left-most cell first.

**Reply by APaganini** [reply]
Ok, many thanks for clarifying this. I think my issue originated from plotting a 1D function `u` using `plt.plot(u.vector())` instead of `plot(u)`.

P.s.: I tried `plt.plot(u.vector())` because `plt.plot(u)` gives an error. I wasn't aware of the distinction between `plt.plot` and `fd.plot`
